### PR TITLE
feat: comply with all parties directions (FPLA-806)

### PR DESCRIPTION
**Pull request by swalker125**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/FPLA-806

### Change description ###

- New event to comply with directions. 
- Story implements functionality to comply with only all parties directions

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: ccd-definition/CaseEvent/PREPARE_FOR_HEARING.json [hunk 1]
```diff
@@ -145,5 +145,47 @@
     "EndButtonLabel": "Save and continue",
     "ShowSummary": "Y",
     "ShowEventNotes": "N"
+  },
+  {
+    "LiveFrom": "01/01/2017",
```
Comment (lewisbirks) [inline ccd-definition/CaseEvent/PREPARE_FOR_HEARING.json]: I'm guessing this duplication was to deal with the issue of not being able to hide the complex type properly

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java [hunk 2]
```diff
@@ -77,21 +82,122 @@
...
+
+    /**
+     * Adds a response by an assignee to a direction where the direction ID of the response matches the direction ID
+     * of the direction.
+     *
+     * @param responses  a list of single responses to directions.
+     * @param directions a list of directions with combined responses.
+     */
+    public void addResponsesToDirections(List<DirectionResponse> responses, List<Element<Direction>> directions) {
+        directions.forEach(direction -> responses.stream()
```
Comment (Gogetter) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: could `directions` and `responses` be null at any point?

File: e2e/tests/localAuthorityCompliesWithDirectionsAfterSDOIsIssued_test.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+const response = require('../fixtures/response');
+const directions = require('../fixtures/directions');
+
+let caseId;
+
+Feature('Comply with directions');
+
+Before(async (I, caseViewPage, submitApplicationEventPage, sendCaseToGatekeeperEventPage, draftStandardDirectionsEventPage) => {
+  if (!caseId) {
+    await I.logInAndCreateCase(config.swanseaLocalAuthorityEmailUserOne, config.localAuthorityPassword);
```
Comment (lewisbirks) [inline e2e/tests/localAuthorityCompliesWithDirectionsAfterSDOIsIssued_test.js]: Maybe extract this, as we get further into the case life cycle we would be repeating steps such as this over and over in new tests

File: ccd-definition/CaseEventToComplexTypes/COMPLY_CAFCASS-cafcassDirections.json [hunk 4]
```diff
@@ -0,0 +1,93 @@
+[
```
Comment (lewisbirks) [inline ccd-definition/CaseEventToComplexTypes/COMPLY_CAFCASS-cafcassDirections.json]: could maybe move these 3 with a comply with directions folder when you merge the changes in that I made the other day

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/DirectionAssignee.java [hunk 5]
```diff
@@ -17,4 +19,11 @@
     public String getValue() {
         return value;
     }
+
+    public static DirectionAssignee fromString(String value) {
+        return Stream.of(DirectionAssignee.values())
+            .filter(x -> x.getValue().equals(value))
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/enums/DirectionAssignee.java]: Does case matter here?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/Direction.java [hunk 6]
```diff
@@ -22,4 +25,14 @@
     private String directionNeeded;
     private String custom;
     private LocalDateTime dateToBeCompletedBy;
+    private DirectionResponse response;
+    private List<Element<DirectionResponse>> responses;
+
+    public List<Element<DirectionResponse>> getResponses() {
+        if (isNull(responses)) {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/Direction.java]: could one line to `return isNull(responses) ? new ArrayList<>() : responses`

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java [hunk 7]
```diff
@@ -77,21 +82,122 @@
...
+    }
+
+    /**
+     * Collects directions for all parties and places them into single map, where the key is the role direction,
+     * and the value is the list of directions associated.
+     *
+     * @param caseData data from case.
+     * @return Map of roles and directions.
+     */
+    public Map<String, List<Element<Direction>>> collectDirectionsToMap(CaseData caseData) {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: Does the map key need to be a `String` can it not be a `DirectionAssignee`? Seems the only place that it is used is in `addHiddenVariablesToResponseForAssignee` where you are then converting the `String` back into a `DirectionAssignee`.

File: e2e/tests/localAuthorityCompliesWithDirectionsAfterSDOIsIssued_test.js (hunk 3 above)
Comment (joez-f) [reply e2e/tests/localAuthorityCompliesWithDirectionsAfterSDOIsIssued_test.js]: good idea - more like tech debt cleanup though; we have duplicate code everywhere in other e2e tests which could probably be extracted

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java [hunk 8]
```diff
@@ -77,21 +82,122 @@
...
+    private List<Element<Direction>> addHiddenVariablesToResponseForAssignee(String assignee,
+                                                                             List<Element<Direction>> directions) {
+        return directions.stream()
+            .filter(element -> isNotEmpty(element.getValue().getResponse()))
+            .filter(element -> isNotEmpty(element.getValue().getResponse().getComplied()))
+            .map(x -> Element.<Direction>builder()
+                .id(x.getId())
+                .value(x.getValue().toBuilder()
+                    .response(x.getValue().getResponse().toBuilder()
+                        .assignee(DirectionAssignee.fromString(assignee))
```
Comment (Gogetter) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: do we mind if `assignee` is `null` here, given `DirectionAssignee.fromString(...)` could return `null`?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java (hunk 2 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: no

File: ccd-definition/CaseEvent/PREPARE_FOR_HEARING.json (hunk 1 above)
Comment (swalker125) [reply ccd-definition/CaseEvent/PREPARE_FOR_HEARING.json]: yep 👍

File: e2e/tests/localAuthorityCompliesWithDirectionsAfterSDOIsIssued_test.js (hunk 3 above)
Comment (swalker125) [reply e2e/tests/localAuthorityCompliesWithDirectionsAfterSDOIsIssued_test.js]: I'm with Joe, I am not too fussed about extracting this logic here. 29 file changes as is 🙁

File: ccd-definition/CaseEventToComplexTypes/COMPLY_CAFCASS-cafcassDirections.json (hunk 4 above)
Comment (swalker125) [reply ccd-definition/CaseEventToComplexTypes/COMPLY_CAFCASS-cafcassDirections.json]: merged your changes, the thing with adding them to comply is that they are different events 🤔 what do you reckon?

File: e2e/tests/localAuthorityCompliesWithDirectionsAfterSDOIsIssued_test.js (hunk 3 above)
Comment (lewisbirks) [reply e2e/tests/localAuthorityCompliesWithDirectionsAfterSDOIsIssued_test.js]: Yeah it was a sort of thing that isn't required but at some point would be nice

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/DirectionAssignee.java (hunk 5 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/enums/DirectionAssignee.java]: so, the keys of our role collections is always the string passed in here. They will always be in the form roleDirections. This will always map to DirectionAssignee

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/Direction.java (hunk 6 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/Direction.java]: I hate inline logic like this. But...
```
 defaultIfNull(responses, new ArrayList<>());
```

is a good shout

File: ccd-definition/CaseEventToComplexTypes/COMPLY_CAFCASS-cafcassDirections.json (hunk 4 above)
Comment (lewisbirks) [reply ccd-definition/CaseEventToComplexTypes/COMPLY_CAFCASS-cafcassDirections.json]: That's a good point but they are all related as well which could be a reason to group them into one folder

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java (hunk 8 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: DirectionAssignee could return null... however we always pass in the keys from our role collections, which will always be the values of DirectionAssignee. So I don't believe this will ever happen

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java (hunk 7 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: this is a good spot, it would actually totally remove the need for my string -> DirectionAssignee method I think 🤔

Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: I think so as well

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java [hunk 9]
```diff
@@ -103,10 +209,35 @@ public void persistHiddenDirectionValues(List<Element<Direction>> directionWithV
...
+    /**
+     * Removes directions where custom is set to Yes.
+     *
+     * @param directions any list of directions.
+     * @return a list of directions that are not custom.
+     */
+    public List<Element<Direction>> removeCustomDirections(List<Element<Direction>> directions) {
+        return directions.stream()
+            .filter(element -> !"Yes".equals(element.getValue().getCustom()))
+            .collect(Collectors.toList());
```
Comment (Gogetter) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: could we have the _Yes_ and _No_ constants in a separate class?

Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: don't see any pros to doing this

**Comment by swalker125** [comment]
TAB VIEW:

Screen Shot 2019-11-08 at 15 12 10

COMPLY YES:

Screen Shot 2019-11-08 at 15 12 27

COMPLY NO:

Screen Shot 2019-11-08 at 15 12 40

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/DirectionAssignee.java (hunk 5 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/enums/DirectionAssignee.java]: method is now removed

File: ccd-definition/ComplexTypes/1_directionResponse.json [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    "ElementLabel": "Has this direction been complied with?",
+    "SecurityClassification": "Public"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "DirectionResponse",
+    "ListElementCode": "file",
+    "FieldType": "Document",
+    "ElementLabel": "Upload file",
+    "HintText": "You can upload JPG, BMP, PNG, TIF or PDF files up to 10MB in size.",
```
Comment (lewisbirks) [inline ccd-definition/ComplexTypes/1_directionResponse.json]: Do we actually make this restriction? Or is it more of a guideline.

Comment (swalker125) [reply ccd-definition/ComplexTypes/1_directionResponse.json]: I think this is GDS standard for upload files... it was in the design, so I added

Comment (lewisbirks) [reply ccd-definition/ComplexTypes/1_directionResponse.json]: I think this [confluence](https://tools.hmcts.net/confluence/display/RCCD/Document+Store+-+Global+Whitelist+of+Document+Types+and+Maximum+File+size) page links to the actual restrictions

Comment (swalker125) [reply ccd-definition/ComplexTypes/1_directionResponse.json]: handy to know! 100MB files 😮

Comment (lewisbirks) [reply ccd-definition/ComplexTypes/1_directionResponse.json]: BIG

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java [hunk 11]
```diff
@@ -77,21 +82,123 @@
...
+            .filter(element -> isNotEmpty(element.getValue().getResponse().getComplied()))
+            .map(x -> Element.<Direction>builder()
+                .id(x.getId())
+                .value(x.getValue().toBuilder()
+                    .response(x.getValue().getResponse().toBuilder()
+                        .assignee(assignee)
+                        .directionId(x.getId())
+                        .build())
+                    .build())
+                .build())
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: That map probably could be extracted. Logic not relevant to what's happening here.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java (hunk 9 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: Yes I actually like this, Constants for EMPTY and SPACE are the most annoying for me ever. 

That Yes too much rooted in our CCD model. But to be honest I think that should be converted to boolean - probably we should have some of our Jackson converter or something like that. Boolean field marked with @YerOrNo annotation, what do you think?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java [hunk 12]
```diff
@@ -103,10 +210,35 @@ public void persistHiddenDirectionValues(List<Element<Direction>> directionWithV
...
+     *
+     * @param directions a list of directions.
+     * @param assignee   the assignee of the directions to be returned.
+     * @return a list of directions belonging to the assignee.
+     */
+    public List<Element<Direction>> getDirectionsForAssignee(List<Element<Direction>> directions,
+                                                             DirectionAssignee assignee) {
+        return directions.stream()
+            .filter(element -> element.getValue().getAssignee().equals(assignee))
+            .collect(Collectors.toList());
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: Nothing major, but it's another instance where we collect just to stream couple of steps later - I think we could be mindful a bit more. What do you think?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperServiceTest.java [hunk 13]
```diff
@@ -29,151 +39,706 @@
...
-        CaseData caseData = populateCaseDataWithFixedDirections()
-            .allPartiesCustom(buildCustomDirections())
-            .localAuthorityDirectionsCustom(buildCustomDirections())
-            .parentsAndRespondentsCustom(buildCustomDirections())
-            .cafcassDirectionsCustom(buildCustomDirections())
-            .otherPartiesDirectionsCustom(buildCustomDirections())
-            .courtDirectionsCustom(buildCustomDirections())
-            .build();
+    @Nested
+    class CombineAllDirectionsTest {
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperServiceTest.java]: If we need to do this maybe it would be worth having separate file? It's a bit difficult to review now, due to all file is marked as changed. Maybe could be refactored as separate PR?

File: ccd-definition/CaseEventToComplexTypes/COMPLY_CAFCASS-cafcassDirections.json (hunk 4 above)
Comment (swalker125) [reply ccd-definition/CaseEventToComplexTypes/COMPLY_CAFCASS-cafcassDirections.json]: hmm so we group by event, technically they are all separate events. Hesitating about putting them in the same folder, so I am going to say no for now. Easy to address in another PR, so I am not too worried about the outcome. Everyone happy?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ComplyWithDirectionsControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+    private List<Direction> directionsForAllRoles() {
+        return ImmutableList.of(
+            Direction.builder().assignee(ALL_PARTIES).build(),
+            Direction.builder().assignee(LOCAL_AUTHORITY).build(),
+            Direction.builder().assignee(PARENTS_AND_RESPONDENTS).build(),
+            Direction.builder().assignee(CAFCASS).build(),
+            Direction.builder().assignee(OTHERS).build(),
+            Direction.builder().assignee(COURT).build()
+        );
+    }
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ComplyWithDirectionsControllerTest.java]: I am going to Stream through the enum values, that's a good idea from @Gogetter. I am not too fussed about extracting more methods. I don't like the idea of nested private methods

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java [hunk 15]
```diff
@@ -77,21 +82,122 @@
...
+        return directions.stream()
+            .map(element -> Element.<Direction>builder()
+                .id(element.getId())
+                .value(element.getValue().toBuilder()
+                    .response(element.getValue().getResponses().stream()
+                        .filter(response -> response.getValue().getDirectionId().equals(element.getId()))
+                        .filter(response -> response.getValue().getAssignee().getValue().equals(assignee))
+                        .map(Element::getValue)
+                        .findFirst()
+                        .orElse(null))
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: empty response in CCD is stored as null. This simply puts the value there if there isn't a response to place in the builder. Can you think of a better implementation?

File: ccd-definition/AuthorisationCaseField/cafcass.json [hunk 16]
```diff
@@ -293,5 +293,19 @@
     "CaseFieldID": "temporaryC2Document",
     "UserRole": "caseworker-publiclaw-cafcass",
     "CRUD": "CRU"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseFieldID": "cafcassDirections",
+    "UserRole": "caseworker-publiclaw-cafcass",
```
[the remark this answers is not in the document]
Comment (swalker125) [reply ccd-definition/AuthorisationCaseField/cafcass.json]: currently discussing with Moray. I have direction from Sara to leave as is for now

File: ccd-definition/AuthorisationCaseEvent/AuthorisationCaseEvent-nonprod.json [hunk 17]
```diff
@@ -19,5 +19,26 @@
   "CaseEventID": "createNoticeOfProceedings-PREPARE_FOR_HEARING",
   "UserRole": "caseworker-publiclaw-courtadmin",
   "CRUD": "CRU"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseEventID": "COMPLY_LOCAL_AUTHORITY",
```
[the remark this answers is not in the document]
Comment (swalker125) [reply ccd-definition/AuthorisationCaseEvent/AuthorisationCaseEvent-nonprod.json]: I'm going to update the description of the events, yes we can add some higher level confluence docs etc. Was that what you had in mind?

[the remark this answers is not in the document]
Comment (swalker125) [reply ccd-definition/AuthorisationCaseEvent/AuthorisationCaseEvent-nonprod.json]:
```
 "Description": "Event gives Local Authority user access to comply with their directions as well as all parties",
```

File: ccd-definition/AuthorisationCaseField/court-admin.json [hunk 18]
```diff
@@ -355,5 +355,12 @@
     "CaseFieldID": "temporaryC2Document",
     "UserRole": "caseworker-publiclaw-courtadmin",
     "CRUD": "CRU"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseFieldID": "courtDirectionsCustom",
```
[the remark this answers is not in the document]
Comment (swalker125) [reply ccd-definition/AuthorisationCaseField/court-admin.json]: added 
```
    "Comment": "courtDirectionsCustom is used here to prevent Gatekeeper inadvertently getting C and D permissions in draftSDO journey on courtDirections. We do not want them to able to create/delete directions"

```

to caseEventToFields

File: ccd-definition/ComplexTypes/2_direction.json [hunk 19]
```diff
@@ -80,5 +80,24 @@
     "ElementLabel": "Status",
     "SecurityClassification": "Public",
     "FieldShowCondition": "directionNeeded!=\"No\""
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "Direction",
+    "ListElementCode": "response",
```
[the remark this answers is not in the document]
Comment (swalker125) [reply ccd-definition/ComplexTypes/2_direction.json]: added comment explaining the UI display nature of response, and the data store nature of responses

File: ccd-definition/FixedLists/C2Uploaded.json [hunk 20]
```diff
@@ -0,0 +1,8 @@
+[
```
[the remark this answers is not in the document]
Comment (swalker125) [reply ccd-definition/FixedLists/C2Uploaded.json]: 😂

File: ccd-definition/FixedLists/YesOrNo.json [hunk 21]
```diff
@@ -0,0 +1,16 @@
+[
```
[the remark this answers is not in the document]
Comment (swalker125) [reply ccd-definition/FixedLists/YesOrNo.json]: hmm, so the only difference here is that fix radio list is displayed vertically, the YesNo field from ccd is horizontal. I will discuss with design I guess?

File: e2e/tests/localAuthorityCompliesWithDirectionsAfterSDOIsIssued_test.js [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+
+Scenario('local authority complies with directions', async (I, caseViewPage, complyWithDirectionsEventPage) => {
+  await I.signIn(config.swanseaLocalAuthorityEmailUserOne, config.localAuthorityPassword);
+  await I.navigateToCaseDetails(caseId);
+  caseViewPage.goToNewActions(config.applicationActions.complyWithDirections);
+  await complyWithDirectionsEventPage.canComplyWithDirection('localAuthorityDirections', 0, response, config.testFile);
+  await I.completeEvent('Save and continue');
+  await I.seeEventSubmissionConfirmation(config.applicationActions.complyWithDirections);
+  caseViewPage.selectTab(caseViewPage.tabs.orders);
+  I.seeAnswerInTab(1, 'Compliance 1', 'Party', 'Local Authority');
```
[the remark this answers is not in the document]
Comment (swalker125) [reply e2e/tests/localAuthorityCompliesWithDirectionsAfterSDOIsIssued_test.js]: hahaha so a hidden field exists at 2 😂

[the remark this answers is not in the document]
Comment (swalker125) [reply e2e/tests/localAuthorityCompliesWithDirectionsAfterSDOIsIssued_test.js]: the way the method works it looks for <tr> tags in the html

File: ccd-definition/FixedLists/YesOrNo.json (hunk 21 above)
[the remark this answers is not in the document]
Comment (swalker125) [reply ccd-definition/FixedLists/YesOrNo.json]: content are fine with horizontal, shall refactor

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ComplyWithDirectionsControllerTest.java [hunk 23]
```diff
@@ -0,0 +1,167 @@
...
+class ComplyWithDirectionsControllerTest {
+
+    @Autowired
+    private MockMvc mockMvc;
+
+    @Autowired
+    private ObjectMapper mapper;
+
+    private static final String AUTH_TOKEN = "Bearer token";
+    private static final String USER_ID = "1";
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ComplyWithDirectionsControllerTest.java]: interestingly enough we used to have a contents file, and it just got out of hand due to abstraction, so we took everything back out. Happy to revisit this idea, but not here :)

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ComplyWithDirectionsControllerTest.java [hunk 24]
```diff
@@ -0,0 +1,167 @@
...
+        AboutToStartOrSubmitCallbackResponse callbackResponse = mapper.readValue(response.getResponse()
+            .getContentAsByteArray(), AboutToStartOrSubmitCallbackResponse.class);
+
+        return mapper.convertValue(callbackResponse.getData(), CaseData.class);
+    }
+
+    private boolean roleDirectionsContainExpectedAllPartiesDirection(List<Element<Direction>> roleDirections) {
+        return roleDirections.stream()
+            .map(Element::getValue)
+            .anyMatch(x -> x.equals(Direction.builder().assignee(ALL_PARTIES).build()));
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ComplyWithDirectionsControllerTest.java]: I have extracted, you're right this probably increases readability :)

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/Direction.java [hunk 25]
```diff
@@ -22,4 +25,10 @@
     private String directionNeeded;
     private String custom;
     private LocalDateTime dateToBeCompletedBy;
+    private DirectionResponse response;
+    private List<Element<DirectionResponse>> responses;
+
+    public List<Element<DirectionResponse>> getResponses() {
+        return defaultIfNull(responses, new ArrayList<>());
+    }
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/Direction.java]: much ealier to deal with when you know you're always getting a list

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/DirectionResponse.java [hunk 26]
```diff
@@ -0,0 +1,24 @@
...
+import uk.gov.hmcts.reform.fpl.enums.DirectionAssignee;
+import uk.gov.hmcts.reform.fpl.model.common.DocumentReference;
+
+import java.util.List;
+import java.util.UUID;
+
+@Data
+@Builder(toBuilder = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class DirectionResponse {
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/DirectionResponse.java]: I will ping a message to Trevor now

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperServiceTest.java [hunk 27]
```diff
@@ -1,24 +1,34 @@
 package uk.gov.hmcts.reform.fpl.service;
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperServiceTest.java]: yes, I've probably gone a bit overboard with refactoring here. I can spend some time reverting this, all though I think it should probably end up at this point eventually...

[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperServiceTest.java]: reverted, still lots of new tests however!

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ComplyWithDirectionsController.java [hunk 28]
```diff
@@ -0,0 +1,88 @@
...
+
+    @PostMapping("/about-to-start")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToStart(@RequestBody CallbackRequest callbackrequest) {
+        CaseDetails caseDetails = callbackrequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        List<Element<Direction>> allPartyDirections = directionHelperService
+            .getDirectionsForAssignee(caseData.getStandardDirectionOrder().getDirections(), ALL_PARTIES);
+
+        directionHelperService.sortDirectionsByAssignee(caseData.getStandardDirectionOrder().getDirections())
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ComplyWithDirectionsController.java]: have extracted to a void method called addAssigneeDirectionKeyValuePairsToCaseData in directionHelperService. Shall add some unit tests 👍

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ComplyWithDirectionsController.java [hunk 29]
```diff
@@ -0,0 +1,88 @@
...
+        CaseDetails caseDetails = callbackrequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        Map<DirectionAssignee, List<Element<Direction>>> directionsMap =
+            directionHelperService.collectDirectionsToMap(caseData);
+
+        List<DirectionResponse> directionsToUpdate = directionHelperService
+            .addHiddenVariablesToResponseForManyAssignees(directionsMap).stream()
+            .map(x -> x.getValue().getResponse())
+            .collect(toList());
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ComplyWithDirectionsController.java]: refactored the helper method to return a List of responses

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java [hunk 30]
```diff
@@ -77,21 +82,123 @@
...
-                        directionToAddValue.getValue().getAssignee(), direction.getValue().getAssignee()));
-
-                    if (!direction.getValue().getReadOnly().equals("No")) {
-                        directionToAddValue.getValue().setDirectionText(direction.getValue().getDirectionText());
-                    }
-                }));
+        directionToAddValues.forEach(directionToAddValue -> directionWithValues
+            .stream()
+            .filter(direction ->
+                direction.getValue().getDirectionType().equals(directionToAddValue.getValue().getDirectionType()))
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: extracted to private hasSameDirectionType method

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java [hunk 31]
```diff
@@ -77,21 +82,123 @@
...
+     *
+     * @param caseData data from case.
+     * @return Map of roles and directions.
+     */
+    public Map<DirectionAssignee, List<Element<Direction>>> collectDirectionsToMap(CaseData caseData) {
+        final List<Element<Direction>> emptyList = ImmutableList.of();
+
+        // courtDirectionsCustom is used due to the workaround in ComplyWithDirectionsController
+        return Map.of(
+            ALL_PARTIES, defaultIfNull(caseData.getAllParties(), emptyList),
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: I like the idea of custom getters, I think I will like to address in a separate PR though. Thoughts?

File: ccd-definition/CaseEventToComplexTypes/COMPLY_CAFCASS-cafcassDirections.json (hunk 4 above)
Comment (lewisbirks) [reply ccd-definition/CaseEventToComplexTypes/COMPLY_CAFCASS-cafcassDirections.json]: I'm fine with that

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java [hunk 32]
```diff
@@ -77,21 +82,123 @@
...
+     * @param directionMap a map of directions where assignee is key and value is a list of directions.
+     * @return a list of directions with hidden variables added where a response is present.
+     */
+    public List<Element<Direction>> addHiddenVariablesToResponseForManyAssignees(
+        Map<DirectionAssignee, List<Element<Direction>>> directionMap) {
+        return directionMap.entrySet()
+            .stream()
+            .map(entry -> addHiddenVariablesToResponseForAssignee(entry.getKey(), entry.getValue()))
+            .flatMap(List::stream)
+            .collect(toList());
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: refactored method now

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java (hunk 11 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: I disagree, the map builds the response. I have moved the filter methods to a private method

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java (hunk 12 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: hmm I want these public methods to be reusable. Hence why I am returning a list so often. It means they can be used anywhere easily, and keeps the methods simple as to what they are doing

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperServiceTest.java (hunk 13 above)
Comment (swalker125) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperServiceTest.java]: I have reverted lots of the changes, as I say, lots of new tests though so I am not sure how helpful it will be 😅

File: e2e/tests/localAuthorityCompliesWithDirectionsAfterSDOIsIssued_test.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (akomisarek) [reply e2e/tests/localAuthorityCompliesWithDirectionsAfterSDOIsIssued_test.js]: 👍

File: ccd-definition/AuthorisationCaseEvent/AuthorisationCaseEvent-nonprod.json (hunk 17 above)
[the remark this answers is not in the document]
Comment (akomisarek) [reply ccd-definition/AuthorisationCaseEvent/AuthorisationCaseEvent-nonprod.json]: Yes. makes sense to me right now. We can create confluence docs as the next step.

File: ccd-definition/FixedLists/YesOrNo.json (hunk 21 above)
[the remark this answers is not in the document]
Comment (akomisarek) [reply ccd-definition/FixedLists/YesOrNo.json]: But that's the interesting point - you get vertical if you use the out of the box YesOrNo, didn't think about this!

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ComplyWithDirectionsControllerTest.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (akomisarek) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ComplyWithDirectionsControllerTest.java]: Of course, not as part of this PR!

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java (hunk 31 above)
[the remark this answers is not in the document]
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: Happy to address in iteration PR - this one is already pretty big.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java (hunk 12 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: What's not reusable in Returning Streams? ;) ok I guess some clients need to collect before doing some operations, but if they need to filter or do other operations, they will need to stream. Happy to discuss this at later point though, definitely don't need to make changes right now! 

https://stackoverflow.com/questions/24676877/should-i-return-a-collection-or-a-stream

The last paragraph from the answers sums my thinking pretty nicely as well:

> So I would say that most of the time, Stream is the right answer -- it is more flexible, it doesn't impose usually-unnecessary materialization costs, and can be easily turned into the Collection of your choice if needed. But sometimes, you may have to return a Collection (say, due to strong consistency requirements), or you may want to return Collection because you know how the user will be using it and know this is the most convenient thing for them.
